cpp/src/heap.cuh

```diff
@@ -0,0 +1,225 @@
+     *
+     */
+    template <typename T, typename S, typename C>
+    inline void __host__ __device__ heapify(T *array, S size, C compare) {
+      static_assert(std::is_integral<S>::value, "Index must be of an integral type");
+
+      //
+      // We want to order ourselves as a heap.  This is accomplished by starting
+      // at the end and for each element, compare with its parent and
+      // swap if necessary.  We repeat this until there are no more swaps
+      // (should take no more than log2(size) iterations).
+      //
+      S count_swaps = 1;
+      while (count_swaps > 0) {
+	count_swaps = 0;
+	for (S i = size - 1 ; i > 0 ; --i) {
+	  S p = detail::parent(i);
+
+	  {
+	    S lc = detail::left_child(p);
+	    S rc = detail::right_child(p);
+	    if (i == lc) {
+	    } else if (i == rc) {
+	    } else {
+	      printf("in heapify... i = %ld, p = %ld, lc = %ld, rc = %ld\n", i, p, lc, rc);
+	    }
+	  }
+	  if (compare(array[i], array[p])) {
+	    thrust::swap(array[i], array[p]);
+	    ++count_swaps;
+	  }
+	}
+      }
+    }
+
+    /**
+     * @brief Pop the top element off of the heap.  Note that the caller
+     *        should decrement the size - the last element in the
+     *        array is no longer used.
+     *
+     * @return - the top of the heap.
+     */
+    template <typename T, typename S, typename C>
+    inline T __host__ __device__ heap_pop(T *array, S size, C compare) {
+      static_assert(std::is_integral<S>::value, "Index must be of an integral type");
+
+      //
+      //  Swap the top of the array with the last element
+      //
+      --size;
+      thrust::swap(array[0], array[size]);
+
+      //
+      //  Now top element is no longer the smallest (largest), so we need
+      //  to sift it down to the proper location.
+      //
+      for (S i = 0 ; i < size ; ) {
+	S lc = detail::left_child(i);
+	S rc = detail::right_child(i);
+	S smaller = i;
+
+	//
+	//  We can go out of bounds, let's check the simple cases
+	//
+	if (rc < size) {
+	  //
+	  //  Both children exist in tree, pick the smaller (lerger)
+	  //  one.
+	  //
+	  smaller = (compare(array[lc], array[rc])) ? lc : rc;
+	} else if (lc < size) {
+	  smaller = lc;
+	}
+
+	if ((smaller != i) && (compare(array[smaller], array[i]))) {
+	  thrust::swap(array[i], array[smaller]);
+	  i = smaller;
+	} else {
+	  //
+	  //  If we don't swap then we can stop checking, break out of the loop
+	  //
+	  i = size;
+	}
+      }
+      
+      return array[size];
+    }
+
+    template <typename T, typename S, typename C>
+    inline void __host__ __device__ check_heap(T *array, S size, C compare) {
+      static_assert(std::is_integral<S>::value, "Index must be of an integral type");
+
+      for (S i = 0 ; i < size ; ++i) {
+	S lc = detail::left_child(i);
+	S rc = detail::right_child(i);
+
+	if ((rc < size) && (compare(array[rc], array[i]))) {
+	  printf("error... array[%ld] = %ld, array[%ld] = %ld, size = %ld\n", rc, array[rc], i, array[i], size);
```

Comment:
Again, are we going to keep this printf?

Reply:
This is a debugging function.  I'm deleting the whole function.